### PR TITLE
[foreach] Use same `dtypes` when  `dtypesIfCUDA` is `None`

### PR DESCRIPTION
**Pull request by crcrpar**
in order to avoid accidentally testing cuda path with fewer dtypes

### CODE REVIEW

**Review by janeyx99 — DISMISSED** [review]
Is the test plan making sure that we test the same number (with a few more) tests compared to before?

Is there an easy way to make sure we haven’t lost coverage?

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 1]
```diff
@@ -388,9 +388,10 @@ void foreach_tensor_zero_cuda_(TensorList tensors) {
   std::vector<std::vector<at::Tensor>> tensor_lists;
   tensor_lists.emplace_back(tensors.vec());
 
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
       ScalarType::Half,
       ScalarType::BFloat16,
+      ScalarType::Bool,
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:394]: Did we miss bool before for foreach_zero?

**Comment by crcrpar** [comment]
`pytest test_foreach.py --collect-only` says there're 4338 tests in this branch and 4278 tests as of 0848051844181a7ac816b555f84caf16f20f0256 which doesn't have #124703 though.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -9544,10 +9557,12 @@ def __call__(self, opinfo, device, dtype, requires_grad, **kwargs):
     ForeachFuncInfo(
         'sin',
         foreach_inputs_sample_func(1, False, False),
+        dtypesIfCUDA=floating_and_complex_types_and(torch.half,),
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:9560]: this means there’s more half support on CUDA vs CPU yes?

**Review by janeyx99 — APPROVED** [review]

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu (hunk 1 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachUnaryOp.cu:394]: yes, it seems to be so.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (crcrpar) [reply torch/testing/_internal/common_methods_invocations.py:9560]: uh, about right I'd say.
I think this is rather historical; when I started this bfloat16 wasn't as popular as today and float16 support on cpu wasn't like today.
I'll try more dtypes in foreach opinfo's (probably with some xfail/skip) after this pr

**Pull request closed without merging**